### DISCUSSION TITLE
upx vs tar.gz

### DISCUSSION THREAD

**Discussion by kuchaguangjie** [upx/upx#679] [General]
In my test, upx produce similar size as tar.gz.
After upx, if I compress it as tar.gz again, the size doesn't reduce, does that means:
1. upx & tar.gz can replace each other, in the sense of size?
2. There is no need to use .tar.gz after ups ?

**Comment by jreiser** [comment]
0. Please give details: file size of input and output, hardware architecture of executable being compressed, any parameters specified when compressing.
1.  tar.gz can contain more than one file at a time; upx cannot.
2. tar.gz doesn't care about the contents of each file. upx requires that a file be executable.
3. Some follow-on software, such as uploads ("Attached files") to github, restricts its inputs to files of certain types (as a precaution against active malware, or other preference), which often do not include ordinary executables, in particular output from upx.  So .gz, .tar.gz, etc., are used as containers for transport.
4. The output from upx is executable; the output from tar.gz is not.

If all that matters is file size, then any reversible compression will do.  If direct execution of the compressed output by the operating system is a requirement, then upx may be the best (or only feasible) solution.

**Reply by kuchaguangjie** [reply]
It's linux mint mate (based on ubuntu 20.04) running on laptop with amd-64 cpu.
Input is a golang binary, of about 15 Mb, output of both upx and tar is about 7.6Mb.

Thanks for the reply, I think I got what I want to know already.
